lib/telemetry_metrics.ex

```diff
@@ -1,5 +1,341 @@
+  application. On each request, you might emit an event with the name of the controller and action
+  handling that request, e.g.:
+
+      Telemetry.execute([:http, :request], 1, %{controller: "user_controller", action: "index"})
+      Telemetry.execute([:http, :request], 1, %{controller: "user_controller", action: "index"})
+      Telemetry.execute([:http, :request], 1, %{controller: "user_controller", action: "create"})
+      Telemetry.execute([:http, :request], 1, %{controller: "product_controller", action: "get"})
+
+  With multi-dimensional data model, the result of aggregating those events by `:controller` and
+  `:action` tags would look like this:
+
+  | controller           | action   | count |
+  |----------------------|----------|-------|
+  | `user_controller`    | `index`  | 2     |
+  | `user_controller`    | `create` | 1     |
+  | `product_controller` | `get`    | 1     |
+
+  You can see that the request count is broken down by unique set of tag values.
+
+  ## Metric types
+
+  Metric type specifies how the event values are aggregated. `Telemetry.Metrics` aims to define
+  a set of metric types covering the most common instrumentation patterns.
+
+  Metric types below are heavily inspired by [OpenCensus](https://opencensus.io).
+
+  ### Counter
+
+  Value of the counter metric is the number of emitted events, regardless of event value. It's
+  monotonically increasing and its value is never reset.
+
+  ### Sum
+
+  Value of the sum metric is the sum of event values.
+
+  ### LastValue
+
+  Value of this metric is the value of the most recent event.
+
+  ### Distribution
+
+  The value of this metric is a histogram distribution of event values, i.e. how many events were
+  emitted with values falling into defined buckets. Histogram values can be used to compute
+  approximation of useful statistics about the data, like quantiles, minimum or maximum.
+
+  ## Metric specifications
+
+  Metric specification is a data structure describing the metric - its name, type, name of the
+  events aggregated by the metric, etc. The structure of metric specification is relevant only to
+  authors of reporters.
+
+  Metric specifications are created using one of the four functions: `counter/2`, `sum/2`,
+  `last_value/2` and `distribution/2`. Each of those functions returns a specification of metric
+  of the corresponding type. The first argument to all these functions is the name of events which
+  are aggregated by the metric. The second argument is a list of options. Below is the description
+  of the options common to all metric types:
+
+  * `:name` - the metric name. Defaults to event name given as first argument;
+  * `:tags` - tags by which aggregations will be broken down. Defaults to an empty list;
+  * `:metadata` - determines what part of event metadata is used as the source of tag values.
+    The default value is `:all`. Note that the specified metadata should contain at least those keys
```

Comment:
I am afraid that `:all` will be a bad default. For example, in Ecto case, we have:

```elixir
%{
      query_time: query_time,
      decode_time: decode_time,
      queue_time: queue_time,
      result: log_result(result),
      params: params,
      query: query_string,
      source: source
    }
```

From those, the result and params are likely too large to include as part of my measurements. So I was thinking it may be better to default to an empty list?

Reply:
Hm, right, this makes more sense.